modules/registry.go

```diff
@@ -12,18 +12,12 @@ var (
 	byName      = make(map[string]interface{})
 )
 
-// RegisterModule registers a module implementation with the specified name.
-// constructor must be a function returning the interface of the module.
+// RegisterModule registers a constructor for a module implementation with the specified name.
 // For example:
 //  RegisterModule("chainedhotstuff", func() consensus.Rules { return chainedhotstuff.New() })
-func RegisterModule(name string, constructor interface{}) {
-	ctorType := reflect.TypeOf(constructor)
-
-	if ctorType.Kind() != reflect.Func && ctorType.NumOut() != 1 && ctorType.Out(0).Kind() != reflect.Interface {
-		panic("invalid argument: constructor must be a function returning an interface")
-	}
-
-	ifaceType := ctorType.Out(0)
+func RegisterModule[T any](name string, constructor func() T) {
```

Comment:
I have a question. Is it useful to allow two or more modules (names) that have the same `T`? I couldn't see any uses that requires more than one module per `T`. In which case, there isn't really a need for the `name` argument, except to allow string-based lookup (via the untyped version). I think the latter could be handled if we assume that the `name` is the same as the package...

Reply:
Would something like this work?
```go
	byName      = make(map[string]any)
)

func RegisterModule[T any](constructor func() T) {
	var t T
	moduleType := reflect.TypeOf(t)
	name := moduleType.PkgPath()
	name = name[strings.LastIndex(name, "/")+1:]

	registryMut.Lock()
	defer registryMut.Unlock()

	byName[name] = constructor
}

func GetModule[T any]() (out T, ok bool) {
	targetType := reflect.TypeOf(out)
	name := targetType.PkgPath()
	name = name[strings.LastIndex(name, "/")+1:]

	registryMut.Lock()
	defer registryMut.Unlock()

	ctor, ok := byName[name]
	if !ok {
		return out, false
	}

	return ctor.(func() T)(), true
}
```

Reply:
Yes, we do want to be able to register multiple modules per type `T`. For example, for the interface type `consensus.Rules` we want to register our three different implementations, `chainedhotstuff`, `fasthotstuff`, and `simplehotstuff`. The registry allows us to construct these implementations at runtime by providing the name of the implementation.